com.unity.multiplayer.mlapi/Runtime/Core/SnapshotSystem.cs

```diff
@@ -0,0 +1,217 @@
+
+                        WriteIndex(buffer);
+                        WriteBuffer(buffer);
+
+                        NetworkManager.Singleton.MessageSender.Send(clientId, NetworkConstants.SNAPSHOT_DATA, NetworkChannel.SnapshotExchange, buffer);
+                        buffer.Dispose();
+
+                    }
+
+                    DebugDisplayStore(m_Snapshot.m_Entries, m_Snapshot.m_LastEntry, "Entries");
+                    DebugDisplayStore(m_Snapshot.m_Entries, m_Snapshot.m_LastEntry, "Received Entries");
+                    break;
+            }
+        }
+
+        private void WriteIndex(NetworkBuffer buffer)
+        {
+            using (var writer = PooledNetworkWriter.Get(buffer))
+            {
+                writer.WriteInt16((short)m_Snapshot.m_LastEntry);
+
+                for (var i = 0; i < m_Snapshot.m_LastEntry; i++)
+                {
+                    writer.WriteUInt64(m_Snapshot.m_Entries[i].m_NetworkObjectId);
+                    writer.WriteUInt16(m_Snapshot.m_Entries[i].m_Index);
+                    writer.WriteUInt16(m_Snapshot.m_Entries[i].m_Position);
+                    writer.WriteUInt16(m_Snapshot.m_Entries[i].m_Length);
+                }
+            }
+        }
+
+        private void WriteBuffer(NetworkBuffer buffer)
+        {
+            // todo: this sends the whole buffer
+            // we'll need to build a per-client list
+            buffer.Write(m_Buffer, 0, m_End);
+        }
+
+        private void AllocateEntry(ref Entry entry, long size)
+        {
+            // todo: deal with free space
+            // todo: deal with full buffer
+
+            entry.m_Position = (ushort)m_Beg;
+            entry.m_Length = (ushort)size;
+            m_Beg += (int)size;
+        }
+
+        public void Store(ulong networkObjectId, int index, INetworkVariable networkVariable)
+        {
+            int pos = m_Snapshot.Find(networkObjectId, index);
+            if (pos == Entry.k_NotFound)
+            {
+                pos = m_Snapshot.AddEntry(networkObjectId, index);
+            }
+
+            // write var into buffer, possibly adjusting entry's position and length
+            using (var varBuffer = PooledNetworkBuffer.Get())
+            {
+                networkVariable.WriteDelta(varBuffer);
+                if (varBuffer.Length > m_Snapshot.m_Entries[pos].m_Length)
+                {
+                    // allocate this Entry's buffer
+                    AllocateEntry(ref m_Snapshot.m_Entries[pos], varBuffer.Length);
+                }
+
+                // Copy the serialized NetworkVariable into our buffer
+                Buffer.BlockCopy(varBuffer.GetBuffer(), 0, m_Buffer, m_Snapshot.m_Entries[pos].m_Position, (int)varBuffer.Length);
+            }
+        }
+
+        public void ReadSnapshot(Stream snapshotStream)
+        {
+            // todo: this is sub-optimal, as it allocates. Review
+            List<int> entriesPositionToRead = new List<int>();
```

Comment:
I assume this will be changed to native list later so it will skip gc and use our native allocator instead.

Reply:
yes. That's why I put the comment. In fact I was seeing to step:
future change 1: important: review that use an allocation free mechanism
future change 2: optional: use a native container